### PR TITLE
Secure Boot support for OUTPUT=USB

### PR DESCRIPTION
**Pull request by jsmeix**
* Type: **Enhancement**

* Impact: **Normal**

* Reference to related issue (URL):

https://github.com/rear/rear/pull/3025#issuecomment-1643774477

* How was this pull request tested?

I tested it same (on same VMs) as I did in
https://github.com/rear/rear/pull/3025#issuecomment-1643774477
but now with the changes here
Secure Boot with OUTPUT=USB works for me.

* Brief description of the changes in this pull request:

In output/USB/Linux-i386/100_create_efiboot.sh
add Secure Boot support for OUTPUT=USB by using
SECURE_BOOT_BOOTLOADER as first stage Secure Boot bootloader BOOTX64.efi
and using grub*.efi as second stage Secure Boot bootloader files
in the same way as already done for OUTPUT=ISO
in output/ISO/Linux-i386/250_populate_efibootimg.sh
see https://github.com/rear/rear/pull/3025#issuecomment-1643774477

My current implementation here is just a first step.
The whole code looks somewhat convoluted
and needs at least some more generic cleanup
to make it easier to further develop things in this area.

### CODE REVIEW

**Comment by jsmeix** [comment]
First things first:
I will not clean up the whole UEFI and Secure Boot code
via this pull request.
Via this pull request only what is needed
to get Secure Boot support with OUTPUT=USB
in the current code environment
should be implemented.

Later - as needed and as time permits - I would like
to clean up step by step the whole UEFI and Secure Boot code.

In particular currently I do not like that via
SECURE_BOOT_BOOTLOADER UEFI_BOOTLOADER is overwritten
because that makes it needlessly hard (at least for me)
to understand the UEFI and Secure Boot code
because it is not clear if in a particular piece of code
UEFI_BOOTLOADER means a non-Secure-Boot bootloader
OR
if UEFI_BOOTLOADER means a Secure Boot first stage bootloader
AND
in the latter case the user cannot configure
the Secure Boot second stage bootloader.
I would like to Keep Separated Items Separated - "KSIS" ;-)

**Comment by jsmeix** [comment]
@pcahyna @rear/contributors 
could you please -as time permits - have a look here?

I would very much appreciate it if you could test it
on non-SUSE Linux distributions, in particular RHEL
and perhaps also Ubuntu and Debian.

**Comment by jsmeix** [comment]
@pcahyna @rear/contributors 
I have a question regarding the
Secure Boot second stage bootloader files:

Currently those are (hardcoded) all files
that match the bash globbing `grub*.efi`
in the directory where SECURE_BOOT_BOOTLOADER is.

I wonder if it is more fail-safe to use all `*.efi` files
in the directory where SECURE_BOOT_BOOTLOADER is,
probably even all `*.efi` files with ignore case matching?

**Comment by jsmeix** [comment]
@pcahyna @rear/contributors 
I would like to merge it tomorrow (Thursday) afternoon
unless there are objections.

**Comment by pcahyna** [comment]
@jsmeix thank you, this is a very important change, sorry for being late with review. I will have a look at it tomorrow.

**Comment by jsmeix** [comment]
@pcahyna 
thank you in advance for your review.
No rush - take your time.
I will wait until you did your review.

**Comment by pcahyna** [comment]
@jsmeix I am a bit late, but looking now. Can you see https://bugzilla.redhat.com/show_bug.cgi?id=2196445#c0 where @rmetrich has provided a quite detailed analysis of what is happening in the current code?

**Comment by jsmeix** [comment]
@pcahyna 
no worries take your time - I will wait until you did your review.

I can access
https://bugzilla.redhat.com/show_bug.cgi?id=2196445#c0

Tomorrow I will try to understand it.
I need time because I am not at all an expert in this area.

**Comment by jsmeix** [comment]
Wow - I am so impressed how fast I am :-))

I think UI do understand the initial decription of @rmetrich 
and - as far as I understand it - this is what I implemented
with this pull request here - at least basically.

@rmetrich 
could you please also have a look here (as time permits).
I would much appreciate it to get a review also from you.

**Comment by jsmeix** [comment]
@pcahyna 
as far as I understand it (which I described in my comments)
```
# Shim is a signed EFI binary that is a first stage bootloader
# that loads and executes another (signed) EFI binary
# which normally is a second stage bootloader
# which normally is a GRUB EFI binary
# which normally is available as a file named grub*.efi
# so when SECURE_BOOT_BOOTLOADER is used as UEFI_BOOTLOADER
# (cf. rescue/default/850_save_sysfs_uefi_vars.sh)
# then Shim (usually shim.efi) must be copied as EFI/BOOT/BOOTX64.efi
# and Shim's second stage bootloader must be also copied where Shim already is.
```
the crucial point for Secure Boot is that
the Shim first stage bootloader "shim.efi" is a signed EFI binary
and Shim's second stage bootloader (usually named grub*.efi)
is also a signed EFI binary
which means
both signed EFI binaries from the Linux distribution
must be used unchanged for Secure Boot
instead of using a unsigned EFI binary
that gets created by build_bootx86_efi()

File: usr/share/rear/output/USB/Linux-i386/100_create_efiboot.sh
Comment (pcahyna) [inline usr/share/rear/output/USB/Linux-i386/100_create_efiboot.sh:68]: Is this code copied from output/ISO/Linux-i386/250_populate_efibootimg.sh ? It looks a bit different (e.g. the `-L` argument to `cp` and different variable names).

Comment (jsmeix) [reply usr/share/rear/output/USB/Linux-i386/100_create_efiboot.sh:68]: I mean:
When we have Secure Boot working
for ISO within the existing ISO code
and for USB within the existing USB code,
then - as a separated second step -
the whole mess could be cleaned up (if possible).

Comment (pcahyna) [reply usr/share/rear/output/USB/Linux-i386/100_create_efiboot.sh:68]: Could you please add a comment explaining whre is the code copied from? Also, would it be feasible to reduce differences between the two versions? For example, why `cp -L` in one version and mere `cp` in other? Or, `test -f "$SECURE_BOOT_BOOTLOADER" || build_bootx86_efi $efi_dst/BOOTX64.efi $efi_dst/grub.cfg "/boot" "$UEFI_BOOTLOADER"` on one version and
```sh
if ! test -f "$SECURE_BOOT_BOOTLOADER" ; then
    build_bootx86_efi $TMP_DIR/mnt/EFI/BOOT/BOOTX64.efi $efi_boot_tmp_dir/grub.cfg "$boot_dir" "$UEFI_BOOTLOADER"
fi
```
in the other.

I am asking for this because if this code needs a cleanup (it does), it will be easier to do if both places to be cleaned up look as similar as possible.

**Comment by jsmeix** [comment]
@pcahyna 
please, first things first:
Does it work for you on RHEL (or Fedora) or not?
If yes, on which RHEL (or Fedora) versions does it work?
If no, where and how does it fail?

@rear/contributors 
it would be nice if someone could test it on Ubuntu or Debian.
But as we do not have an active Ubuntu or Debian maintainer
those distributions are of rather low interest for me.

File: usr/share/rear/output/USB/Linux-i386/100_create_efiboot.sh
Comment (jsmeix) [reply usr/share/rear/output/USB/Linux-i386/100_create_efiboot.sh:68]: @pcahyna 
basically you ask me for some cleanup right now.
I will have a look if I could do something right now.

`cp -L` was already there in the existing USB code
so I used it there also for my new code.

`cp -L` was not used in the existing ISO code
so I did not introduce it there.

On first glance `cp -L` looks better than plain `cp`
but I don't know the reason behind why it is different
and right now I do not like to investigate about this.
This belongs to a later cleanup of that whole mess.

Comment (jsmeix) [reply usr/share/rear/output/USB/Linux-i386/100_create_efiboot.sh:78]: @pcahyna 
with
https://github.com/rear/rear/commit/efb37fb9b535c9d04d97d9711cc05945972ced8f
I did what you asked for except the `cp -L` cleanup
(e.g. using `cp -L` also in 250_populate_efibootimg.sh).

**Comment by pcahyna** [comment]
> @pcahyna please, first things first: Does it work for you on RHEL (and/or Fedora) or not? If yes, on which RHEL (and/or Fedora) versions does it work? If no, where and how does it fail?

I have not tried it yet - I merely looked at the code. I can schedule a test on non-secureboot hardware to make sure that the change does not break the case that was working before. I am interested to test with Secure Boot as well, but this will have to wait a bit as I don't have a test environment set up (I will probably be able to do it over the weekend).

Regarding versions: Fedora 37 and later probably won't work because of #2971 (same problem with RHEL 9), although I could probably work around this by configuring ReaR to use Secure Boot shim even if the firmware is not actually set up for Secure Boot.

**Comment by jsmeix** [comment]
On non-secureboot hardware
SECURE_BOOT_BOOTLOADER must not be specified and
then there should be no change in what gets done
(because of the `if ... $SECURE_BOOT_BOOTLOADER`).

**Comment by pcahyna** [comment]
> On non-secureboot hardware SECURE_BOOT_BOOTLOADER must not be specified

Why is that the case? The shim bootloader is installed even on non-secureboot hardware, shouldn't it just work even without secureboot enabled?

**Comment by jsmeix** [comment]
Because SECURE_BOOT_BOOTLOADER is nowhere set to a non-empty value.
Currently there is no Secure Boot autodetection code in ReaR
as far as I understand things in this code area
where I am not yet familiar with.

**Comment by jsmeix** [comment]
Of course Secure Boot autodetection in ReaR
would be an "expected" feature.

But this is nothing that belongs to this pull request
and it is even nothing that belongs to the subsequent
cleanup of the whole UEFI booting setup code in ReaR
(because 'cleanup' != 'new features').

So I think Secure Boot autodetection in ReaR
should not be impemented as long as we still
have some mess of UEFI booting setup code
but after we properly cleaned up our current code
to make further development more fail-safe.

General improvements of UEFI related things in ReaR
are on my longer term TODO list.

What I am currently doing here is trying to move
step by step and carefully into this direction.

By the way:
Didn't I somewhere already mention "first things first"?
;-))

**Comment by jsmeix** [comment]
Now it's weekend time!
I wish you all a relaxed and recovering weekend!

**Comment by pcahyna** [comment]
I see only one minor problem: during boot I get
```
error: ../../grub-core/fs/fshelp.c:258:file `/EFI/redhat/x86_64-efi/xzio.mod'
 not found.
 Loading kernel /EFI/BOOT/kernel ...
 Loading initial ramdisk /EFI/BOOT/initrd.cgz ...
 
 Press any key to continue...
```
because the menu contains
```
    insmod xzio
```
and xzio is apparently not present in our default GRUB image. This does not seem to cause any harm and if I don't press any key, the boot continues in about 10 seconds. I wonder whether the `insmod xzio` part can be safely removed (it won't work in any case when Secure Boot is turned on, AIUI).

File: usr/share/rear/output/USB/Linux-i386/100_create_efiboot.sh
Comment (pcahyna) [reply usr/share/rear/output/USB/Linux-i386/100_create_efiboot.sh:78]: @jsmeix thanks, it looks much clearer now

**Review by pcahyna — APPROVED** [review]

**Comment by pcahyna** [comment]
@jsmeix 
> I have a question regarding the Secure Boot second stage bootloader files:
> 
> Currently those are (hardcoded) all files that match the bash globbing `grub*.efi` in the directory where SECURE_BOOT_BOOTLOADER is.
> 
> I wonder if it is more fail-safe to use all `*.efi` files in the directory where SECURE_BOOT_BOOTLOADER is, probably even all `*.efi` files with ignore case matching?

No. I would do what `OUTPUT=ISO` is doing, unless there are known problems with it. If you copy all `*.efi` files, you may also copy `fbx64.efi` (https://www.rodsbooks.com/efi-bootloaders/fallback.html), and that's probably not what one wants.

**Comment by jsmeix** [comment]
@pcahyna 
many thanks for your review and your explanatory comments!
It is much appeciated!

I think I also experienced the same as you wrote in your
https://github.com/rear/rear/pull/3031#issuecomment-1668061462
As it "somehow works" we should leave it "as is" for now
and focus more on the general cleanup of UEFI booting.

Regarding your
https://github.com/rear/rear/pull/3031#issuecomment-1668096663

I think also SUSE always boots via Shim on UEFI systems
(i.e. Shim = fist stage, GRUB2 = second stage) and
I was also wondering if ReaR should for UEFI booting
always "just copy" the (signed) EFI binaries of the
Linux distribution (i.e. shim*.efi and grub*.efi)
instead of making its own ReaR specific EFI binaries.

My current offhanded reasoning:

When the EFI binaries of the Linux distribution
make the original system boot,
then exactly those EFI binaries
should also make the ReaR recovery system boot
on replacement hardware
because replacement hardware must be sufficiently
compatible with the original system hardware.

Another (hopefully reasonable) assumption is that
the (signed) EFI binaries of the Linux distribution
should be sufficiently feature complete
to boot the original system on various hardware
from various kind of boot media (disk, CDROM, ISO).
I assume Linux distribution use exactly the same
(signed) EFI binaries on their installation media
to boot their installation systems and
also later on the disk of the installed system
to boot the installed system.
So I think exactly those (signed) EFI binaries
should also make the ReaR recovery system boot
on sufficiently compatible replacement hardware
from various kind of boot media.
Again there is the requirement that the replacement hardware
is sufficiently compatible with the original system hardware
which means in particular that the ReaR recovery system
must be booted from a medium that is a supported medium
to install the original system.

File: usr/share/rear/output/USB/Linux-i386/100_create_efiboot.sh
Comment (jsmeix) [reply usr/share/rear/output/USB/Linux-i386/100_create_efiboot.sh:78]: @pcahyna 
my pleasure!

**Comment by jsmeix** [comment]
@pcahyna 
regarding your
https://github.com/rear/rear/pull/3031#issuecomment-1668103656

My "secret final plan" is to have
some reasonable UEFI boot autodetection
so UEFI_BOOTLOADER and SECURE_BOOT_BOOTLOADER
do not need to be specified by the user
but the user can use them if needed to specify
the first stage Secure Boot bootloader EFI binary
via SECURE_BOOT_BOOTLOADER and
the second stage Secure Boot bootloader EFI binary
via UEFI_BOOTLOADER
or - when only UEFI_BOOTLOADER is specified
the single stage UEFI bootloader EFI binary.

**Comment by jsmeix** [comment]
@rear/contributors 
I will merge it today afternoon unless objections appear.

**Comment by pcahyna** [comment]
@jsmeix thanks a lot for this improvement, in the meantime I tried it on a Secure Boot enabled hardware with RHEL 8 (it works - both the rescue image an the recovered system) and on Secure Boot disabled (but with configured Secure Boot in ReaR via `SECURE_BOOT_BOOTLOADER=/boot/efi/EFI/redhat/shimx64.efi`)  with RHEL 9 (I don't test Secure Boot this way, but it confirms that using the shim + the default GRUB avoids https://github.com/rear/rear/issues/2971).

**Comment by pcahyna** [comment]
> Another (hopefully reasonable) assumption is that
> the (signed) EFI binaries of the Linux distribution
> should be sufficiently feature complete
> to boot the original system on various hardware
> from various kind of boot media (disk, CDROM, ISO).

I believe that with Secure Boot, you are not allowed to insmod any GRUB modules, so this should be a reasonable assumption - the GRUB image should be indeed feature complete, avoiding the need for grub2-mkstandalone (it also does not need to have the path to the configuration file hardcoded, unlike on PC BIOS, because EFI passes to GRUB the directory it was loaded from and GRUB locates and loads grub.cfg in the same directory).

**Comment by pcahyna** [comment]
> When the EFI binaries of the Linux distribution
> make the original system boot,
> then exactly those EFI binaries
> should also make the ReaR recovery system boot
> on replacement hardware
> because replacement hardware must be sufficiently
> compatible with the original system hardware.

It is not only the recovery (rescue) system (where some deviations may be necessary), it is also the recovered system. In my experience, when recovering without `SECURE_BOOT_BOOTLOADER` set, the recovered system has a UEFI entry of this type
```
Boot0016* RedHatEnterpriseServer 8      HD(1,GPT,f18ee8ed-2fdd-4984-8e71-95a82f99318d,0x800,0x12c000)/File(\EFI\redhat\grubx64.efi)
```
so, GRUB is directly booted, in contrast with the original system, which has
```
Boot000E* Red Hat Enterprise Linux      HD(1,GPT,a1d3cffa-2eb7-4a8a-b559-7f2241cba776,0x800,0x12c000)/File(\EFI\redhat\shimx64.efi)
```
(as it uses the shim always) - so the recovered system is different in this respect from the original system. This will be a problem if the user decides to turn Secure Boot on after the system was backed up and recovered.

> My "secret final plan" is to have
> some reasonable UEFI boot autodetection
> so UEFI_BOOTLOADER and SECURE_BOOT_BOOTLOADER
> do not need to be specified by the user
> but the user can use them if needed to specify
> the first stage Secure Boot bootloader EFI binary
> via SECURE_BOOT_BOOTLOADER and
> the second stage Secure Boot bootloader EFI binary
> via UEFI_BOOTLOADER
> or - when only UEFI_BOOTLOADER is specified
> the single stage UEFI bootloader EFI binary.

regarding autodetection, I believe it is needed to examine the current boot entry (BootCurrent in efibootmgr output) during mkrescue and use it to set the variables - something that is AFAIK not being done now.

**Comment by pcahyna** [comment]
FYI that's what's inside the system ESP on RHEL:
```
# ls -lR /boot/efi/EFI/
/boot/efi/EFI/:
total 8
drwx------. 2 root root 4096 Aug  8 00:03 BOOT
drwx------. 3 root root 4096 Aug  8 05:21 redhat

/boot/efi/EFI/BOOT:
total 1016
-rwx------. 1 root root 943520 Jun  7  2022 BOOTX64.EFI
-rwx------. 1 root root  90568 Jun  7  2022 fbx64.efi

/boot/efi/EFI/redhat:
total 4872
-rwx------. 1 root root     182 Jun  7  2022 BOOTX64.CSV
drwx------. 2 root root    4096 Feb  6 16:27 fonts
-rwx------. 1 root root    6201 Aug  8 05:21 grub.cfg
-rwx------. 1 root root    6201 Aug  8 05:21 grub.cfg.rearbak
-rwx------. 1 root root    1024 Aug  8 00:07 grubenv
-rwx------. 1 root root 2217416 Feb  6 16:27 grubx64.efi
-rwx------. 1 root root  852552 Jun  7  2022 mmx64.efi
-rwx------. 1 root root  943520 Jun  7  2022 shimx64.efi
-rwx------. 1 root root  936520 Jun  7  2022 shimx64-redhat.efi

/boot/efi/EFI/redhat/fonts:
total 0
```
and what's on the ESP of the ReaR bootable disk after your change, when `SECURE_BOOT_BOOTLOADER=/boot/efi/EFI/redhat/shimx64.efi`
```
# ls -lR /mnt/rear/EFI/
/mnt/rear/EFI/:
total 4
drwxr-xr-x. 2 root root 4096 Aug  8 05:05 BOOT

/mnt/rear/EFI/BOOT:
total 540420
-rwxr-xr-x. 1 root root    943520 Aug  8 05:02 BOOTX64.efi
-rwxr-xr-x. 1 root root      1411 Aug  8 05:05 grub.cfg
-rwxr-xr-x. 1 root root   2217416 Aug  8 05:02 grubx64.efi
-rwxr-xr-x. 1 root root 539374222 Aug  8 05:02 initrd.cgz
-rwxr-xr-x. 1 root root  10838352 Aug  8 05:02 kernel
```

**Comment by pcahyna** [comment]
> @pcahyna regarding your [#3031 (comment)](https://github.com/rear/rear/pull/3031#issuecomment-1668103656)
> 
> My "secret final plan" is to have some reasonable UEFI boot autodetection so UEFI_BOOTLOADER and SECURE_BOOT_BOOTLOADER do not need to be specified by the user but the user can use them if needed to specify the first stage Secure Boot bootloader EFI binary via SECURE_BOOT_BOOTLOADER and the second stage Secure Boot bootloader EFI binary via UEFI_BOOTLOADER or - when only UEFI_BOOTLOADER is specified the single stage UEFI bootloader EFI binary.

@jsmeix I have been working on this "secret final plan" and intend to send a PR this week.

In the meantime, since I test only on Fedora-derived distros, I think it could be useful to have the following information for openSUSE derived distros (and any other):
- content of the ESP `ls -lR /boot/efi/EFI`
- content of the `boot.csv` or `BOOT.CSV` file if any ( `iconv -f UTF-16 < /boot/efi/EFI/opensuse/boot.csv` or whatever )
- if there is `grub.efi` or `shim.efi` file, whether it is 32-bit or 64-bit (`file` will tell you).

**Comment by jsmeix** [comment]
@pcahyna 
on my openSUSE Leap 15.5 system:

content of the ESP:
```
# ls -lR /boot/efi/EFI

/boot/efi/EFI:
total 16
drwxr-xr-x 2 root root 8192 Feb 14 12:54 boot
drwxr-xr-x 2 root root 8192 Feb 14 12:54 opensuse

/boot/efi/EFI/boot:
total 1872
-rwxr-xr-x 1 root root 852408 Feb 26 07:23 MokManager.efi
-rwxr-xr-x 1 root root 953800 Feb 26 07:23 bootx64.efi
-rwxr-xr-x 1 root root  90592 Feb 26 07:23 fallback.efi

/boot/efi/EFI/opensuse:
total 3264
-rwxr-xr-x 1 root root  852408 Feb 26 07:23 MokManager.efi
-rwxr-xr-x 1 root root      58 Feb 26 07:23 boot.csv
-rwxr-xr-x 1 root root     173 Feb 26 07:23 grub.cfg
-rwxr-xr-x 1 root root 1275904 Feb 26 07:23 grub.efi
-rwxr-xr-x 1 root root  225280 Feb 26 07:23 grubx64.efi
-rwxr-xr-x 1 root root  953800 Feb 26 07:23 shim.efi
```

content of the boot*.csv or BOOT*.CSV file if any
```
# iconv -f UTF-16 < /boot/efi/EFI/opensuse/boot.csv
shim.efi,opensuse-secureboot
```

grub.efi or shim.efi file, whether it is 32-bit or 64-bit
```
# find /boot/efi/EFI -type f -name '*.efi' | xargs file

/boot/efi/EFI/boot/bootx64.efi:        PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/boot/efi/EFI/boot/fallback.efi:       PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/boot/efi/EFI/boot/MokManager.efi:     PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/boot/efi/EFI/opensuse/MokManager.efi: PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/boot/efi/EFI/opensuse/grub.efi:       PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/boot/efi/EFI/opensuse/shim.efi:       PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
/boot/efi/EFI/opensuse/grubx64.efi:    PE32+ executable (EFI application) x86-64 (stripped to external PDB), for MS Windows
```

**Pull request merged**
